### PR TITLE
prevent opaque types from appearing in impl headers

### LINKED ISSUE #84660: Types defined with type_alias_impl_trait can be used for dispatch in trait impls. Which is unsound.
Let’s define a type:
```rust
trait Foo {}
impl Foo for () {}
type Bar = impl Foo;
fn _defining_use() -> Bar {}
```

Well, a straightforward trait implementation does not work:
```rust
trait Trait {}

impl Trait for Bar {}
```
```
   Compiling playground v0.0.1 (/playground)
error: cannot implement trait on type alias impl trait
  --> src/lib.rs:10:1
   |
10 | impl Trait for Bar {}
   | ^^^^^^^^^^^^^^^^^^
   |
note: type alias impl trait defined here
  --> src/lib.rs:5:12
   |
5  | type Bar = impl Foo;
   |            ^^^^^^^^

error: aborting due to previous error
```
([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=ca0203907352dece29a4645be0e5bfba))

But `rustc` doesn’t complain about anything slighty more complex at the moment, e.g.
```rust
trait Trait {}
impl Trait for (Bar,) {}

trait TraitArg<T> {}
impl TraitArg<Bar> for () {}
```
([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=0db47061b75d65330102c6b869e74132))

Which turns out to be problematic:
```rust
trait TraitArg<T> {
    fn f();
}

impl TraitArg<Bar> for () {
    fn f() {
        println!("ho");
    }
}

fn main() {
    <() as TraitArg<Bar>>::f();
}
```
```
   Compiling playground v0.0.1 (/playground)
thread 'rustc' panicked at 'called `Result::unwrap()` on an `Err` value: ErrorReported', compiler/rustc_mir/src/monomorphize/collector.rs:827:84
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.53.0-nightly (727d10156 2021-04-27) running on x86_64-unknown-linux-gnu

note: compiler flags: -C embed-bitcode=no -C codegen-units=1 -C debuginfo=2 --crate-type bin

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [collect_and_partition_mono_items] collect_and_partition_mono_items
end of query stack
error: internal compiler error: Encountered error `Unimplemented` selecting `Binder(<() as TraitArg<()>>, [])` during codegen
  |
  = note: delayed at compiler/rustc_trait_selection/src/traits/codegen.rs:68:32

thread 'rustc' panicked at 'no errors encountered even though `delay_span_bug` issued', compiler/rustc_errors/src/lib.rs:1018:13
stack backtrace:
   0:     0x7f995e14cfd0 - std::backtrace_rs::backtrace::libunwind::trace::hdcf4f90f85129e83
                               at /rustc/727d101561f9b1e81c6282943292d990288ca479/library/std/src/../../backtrace/src/backtrace/libunwind.rs:90:5
   1:     0x7f995e14cfd0 - std::backtrace_rs::backtrace::trace_unsynchronized::h2669e30cb82f6732
                               at /rustc/727d101561f9b1e81c6282943292d990288ca479/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f995e14cfd0 - std::sys_common::backtrace::_print_fmt::hfbda19e17f6db318
                               at /rustc/727d101561f9b1e81c6282943292d990288ca479/library/std/src/sys_common/backtrace.rs:67:5
   3:     0x7f995e14cfd0 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h1a8751bf59281272
                               at /rustc/727d101561f9b1e81c6282943292d990288ca479/library/std/src/sys_common/backtrace.rs:46:22
   4:     0x7f995e1be31f - core::fmt::write::h7aa6cd0067dca82a
                               at /rustc/727d101561f9b1e81c6282943292d990288ca479/library/core/src/fmt/mod.rs:1094:17
   5:     0x7f995e1418d5 - std::io::Write::write_fmt::hd7dd3a1df9b6befb
                               at /rustc/727d101561f9b1e81c6282943292d990288ca479/library/std/src/io/mod.rs:1584:15
   6:     0x7f995e150d2b - std::sys_common::backtrace::_print::h551e9ec8a9fa8106
                               at /rustc/727d101561f9b1e81c6282943292d990288ca479/library/std/src/sys_common/backtrace.rs:49:5
   7:     0x7f995e150d2b - std::sys_common::backtrace::print::ha4b1c5e95fa040b3
                               at /rustc/727d101561f9b1e81c6282943292d990288ca479/library/std/src/sys_common/backtrace.rs:36:9
   8:     0x7f995e150d2b - std::panicking::default_hook::{{closure}}::h0b34c9ab7fb9f857
                               at /rustc/727d101561f9b1e81c6282943292d990288ca479/library/std/src/panicking.rs:208:50
   9:     0x7f995e15080d - std::panicking::default_hook::h3067e8318decd17a
                               at /rustc/727d101561f9b1e81c6282943292d990288ca479/library/std/src/panicking.rs:225:9
  10:     0x7f995e91382d - rustc_driver::report_ice::hf065daa4796f245b
  11:     0x7f995e151440 - std::panicking::rust_panic_with_hook::h81b8facc50f34daa
                               at /rustc/727d101561f9b1e81c6282943292d990288ca479/library/std/src/panicking.rs:595:17
  12:     0x7f995e151017 - std::panicking::begin_panic_handler::{{closure}}::ha376ab85d95a000e
                               at /rustc/727d101561f9b1e81c6282943292d990288ca479/library/std/src/panicking.rs:497:13
  13:     0x7f995e14d48c - std::sys_common::backtrace::__rust_end_short_backtrace::h6795c8afdd1a77e6
                               at /rustc/727d101561f9b1e81c6282943292d990288ca479/library/std/src/sys_common/backtrace.rs:141:18
  14:     0x7f995e150f79 - rust_begin_unwind
                               at /rustc/727d101561f9b1e81c6282943292d990288ca479/library/std/src/panicking.rs:493:5
  15:     0x7f995e11cdbb - std::panicking::begin_panic_fmt::hf43a0025042538e2
                               at /rustc/727d101561f9b1e81c6282943292d990288ca479/library/std/src/panicking.rs:435:5
  16:     0x7f9960d9e029 - rustc_errors::HandlerInner::flush_delayed::h340bd563bba016cc
  17:     0x7f9960d9cb1b - <rustc_errors::HandlerInner as core::ops::drop::Drop>::drop::h56e016ed68f55f61
  18:     0x7f9960406cd6 - core::ptr::drop_in_place<rustc_session::parse::ParseSess>::hf9273749d5982cb2
  19:     0x7f99604157e1 - <alloc::rc::Rc<T> as core::ops::drop::Drop>::drop::h2bbea2bbb380f5f7
  20:     0x7f99603fba0d - core::ptr::drop_in_place<rustc_interface::interface::Compiler>::h01f06b96a4e05fdc
  21:     0x7f99603fb5e4 - rustc_span::with_source_map::hafa0f290d807fcb7
  22:     0x7f99603fd161 - rustc_interface::interface::create_compiler_and_run::h31928174eaa223b1
  23:     0x7f99604168e8 - scoped_tls::ScopedKey<T>::set::h332c2cb9377dc4c6
  24:     0x7f9960416e03 - std::sys_common::backtrace::__rust_begin_short_backtrace::h3f92abbf0f820c98
  25:     0x7f996041a625 - core::ops::function::FnOnce::call_once{{vtable.shim}}::h3affa87e0b1d695f
  26:     0x7f995e15fe57 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h3aa31cb6360b59d9
                               at /rustc/727d101561f9b1e81c6282943292d990288ca479/library/alloc/src/boxed.rs:1546:9
  27:     0x7f995e15fe57 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h7719d3c7c5841461
                               at /rustc/727d101561f9b1e81c6282943292d990288ca479/library/alloc/src/boxed.rs:1546:9
  28:     0x7f995e15fe57 - std::sys::unix::thread::Thread::new::thread_start::hfbe13ead469fd0bc
                               at /rustc/727d101561f9b1e81c6282943292d990288ca479/library/std/src/sys/unix/thread.rs:71:17
  29:     0x7f995e09c609 - start_thread
  30:     0x7f995dfb0293 - clone
  31:                0x0 - <unknown>

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.53.0-nightly (727d10156 2021-04-27) running on x86_64-unknown-linux-gnu

note: compiler flags: -C embed-bitcode=no -C codegen-units=1 -C debuginfo=2 --crate-type bin

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
thread panicked while panicking. aborting.
error: could not compile `playground`
```
([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=a941c9fbeb5bebb653c2dba7d93f7f99))

<hr>

Unsurprisingly, this is actually not only resulting in ICEs but also in unsoundness:

```rust
#![feature(min_type_alias_impl_trait)]

trait Foo {}
impl Foo for () {}
type Bar = impl Foo;
fn _defining_use() -> Bar {}

trait Trait<T, In> {
    type Out;
    fn convert(i: In) -> Self::Out;
}

impl<In, Out> Trait<Bar, In> for Out {
    type Out = Out;
    fn convert(_i: In) -> Self::Out {
        unreachable!();
    }
}

impl<In, Out> Trait<(), In> for Out {
    type Out = In;
    fn convert(i: In) -> Self::Out {
        i
    }
}

fn transmute<In, Out>(i: In) -> Out {
    <Out as Trait<Bar, In>>::convert(i)
}

fn main() {
    let d;
    {
        let x = "Hello World".to_string();
        d = transmute::<&String, &String>(&x);
    }
    println!("{}", d);
}
```
```
@U
```
([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=676c3dd309b31ec55b89ed5ba011de0c))

@rustbot modify labels: A-typesystem, A-traits, A-impl-trait, F-type_alias_impl_trait, T-compiler, requires-nightly
and someone please add “I-unsound 💥”.

### LINKED ISSUE #86411: Incoherent impls are still allowed on opaque types
I tried this code (modified version of #74298):

```rust
#![feature(min_type_alias_impl_trait)]

type X<T> = impl Sized;

fn f<T>() -> X<T> {}

trait Y {
    fn g(&self) {}
}

impl Y for (X<()>,) {}
impl Y for (X<i32>,) {}

fn main() {
    (f::<()>(),).g();
}
```

I expected to see this happen: incoherent impls error

Instead, this happened: ICE during codegen

### Meta

`rustc --version --verbose`:
```
rustc 1.54.0-nightly (657bc0188 2021-05-31)
binary: rustc
commit-hash: 657bc01888e6297257655585f9c475a0801db6d2
commit-date: 2021-05-31
host: x86_64-unknown-linux-gnu
release: 1.54.0-nightly
LLVM version: 12.0.1
```

<details><summary>Backtrace</summary>
<p>

```

thread 'rustc' panicked at 'no errors encountered even though `delay_span_bug` issued', compiler/rustc_errors/src/lib.rs:1021:13
stack backtrace:
   0:     0x7f8294e6af80 - std::backtrace_rs::backtrace::libunwind::trace::h1a39bd9a98540471
                               at /rustc/657bc01888e6297257655585f9c475a0801db6d2/library/std/src/../../backtrace/src/backtrace/libunwind.rs:90:5
   1:     0x7f8294e6af80 - std::backtrace_rs::backtrace::trace_unsynchronized::h21a1eeae7103ab3f
                               at /rustc/657bc01888e6297257655585f9c475a0801db6d2/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f8294e6af80 - std::sys_common::backtrace::_print_fmt::hc9ffbae9ed6a9871
                               at /rustc/657bc01888e6297257655585f9c475a0801db6d2/library/std/src/sys_common/backtrace.rs:67:5
   3:     0x7f8294e6af80 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h17510753a34a3f09
                               at /rustc/657bc01888e6297257655585f9c475a0801db6d2/library/std/src/sys_common/backtrace.rs:46:22
   4:     0x7f8294ed89ac - core::fmt::write::h1e5a1f350e43b10f
                               at /rustc/657bc01888e6297257655585f9c475a0801db6d2/library/core/src/fmt/mod.rs:1110:17
   5:     0x7f8294e5c805 - std::io::Write::write_fmt::h06ec27c6d028baf1
                               at /rustc/657bc01888e6297257655585f9c475a0801db6d2/library/std/src/io/mod.rs:1588:15
   6:     0x7f8294e6eddb - std::sys_common::backtrace::_print::h1a02603349b1dc60
                               at /rustc/657bc01888e6297257655585f9c475a0801db6d2/library/std/src/sys_common/backtrace.rs:49:5
   7:     0x7f8294e6eddb - std::sys_common::backtrace::print::h0fc317d31c48cd9b
                               at /rustc/657bc01888e6297257655585f9c475a0801db6d2/library/std/src/sys_common/backtrace.rs:36:9
   8:     0x7f8294e6eddb - std::panicking::default_hook::{{closure}}::ha522601c22d7057b
                               at /rustc/657bc01888e6297257655585f9c475a0801db6d2/library/std/src/panicking.rs:208:50
   9:     0x7f8294e6e8b1 - std::panicking::default_hook::hd50ab173af9a3ce8
                               at /rustc/657bc01888e6297257655585f9c475a0801db6d2/library/std/src/panicking.rs:225:9
  10:     0x7f829563f78d - rustc_driver::report_ice::h0bbc0ae77d597594
  11:     0x7f8294e6f5e6 - std::panicking::rust_panic_with_hook::hb5a01416e02405ad
                               at /rustc/657bc01888e6297257655585f9c475a0801db6d2/library/std/src/panicking.rs:626:17
  12:     0x7f8294e6f0c7 - std::panicking::begin_panic_handler::{{closure}}::hf1dda5d5c0706ee0
                               at /rustc/657bc01888e6297257655585f9c475a0801db6d2/library/std/src/panicking.rs:519:13
  13:     0x7f8294e6b45c - std::sys_common::backtrace::__rust_end_short_backtrace::h56c02e9609085b17
                               at /rustc/657bc01888e6297257655585f9c475a0801db6d2/library/std/src/sys_common/backtrace.rs:141:18
  14:     0x7f8294e6f029 - rust_begin_unwind
                               at /rustc/657bc01888e6297257655585f9c475a0801db6d2/library/std/src/panicking.rs:515:5
  15:     0x7f8294e37eeb - std::panicking::begin_panic_fmt::h5889dddec113b3e8
                               at /rustc/657bc01888e6297257655585f9c475a0801db6d2/library/std/src/panicking.rs:457:5
  16:     0x7f8297ac5969 - rustc_errors::HandlerInner::flush_delayed::h71762d6b78304836
  17:     0x7f8297ac441d - <rustc_errors::HandlerInner as core::ops::drop::Drop>::drop::h4ee3903f55419f69
  18:     0x7f8297140d06 - core::ptr::drop_in_place<rustc_session::parse::ParseSess>::hd706f063b3ef96ed
  19:     0x7f829714334e - <alloc::rc::Rc<T> as core::ops::drop::Drop>::drop::h30ee59f057910eb2
  20:     0x7f82971464dd - core::ptr::drop_in_place<rustc_interface::interface::Compiler>::hf2f304d3f5096f94
  21:     0x7f8297145f50 - rustc_span::with_source_map::h6ee13da35fa72295
  22:     0x7f8297148ac8 - scoped_tls::ScopedKey<T>::set::h9c6565b722913393
  23:     0x7f829716d9db - std::sys_common::backtrace::__rust_begin_short_backtrace::hdcf17779193f6d57
  24:     0x7f82971447b5 - core::ops::function::FnOnce::call_once{{vtable.shim}}::hf48084e1b458d4b2
  25:     0x7f8294e7bad7 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hf00ff0dfc0b9e39f
                               at /rustc/657bc01888e6297257655585f9c475a0801db6d2/library/alloc/src/boxed.rs:1575:9
  26:     0x7f8294e7bad7 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h236979b1b61b3517
                               at /rustc/657bc01888e6297257655585f9c475a0801db6d2/library/alloc/src/boxed.rs:1575:9
  27:     0x7f8294e7bad7 - std::sys::unix::thread::Thread::new::thread_start::h8f357e546e4775fa
                               at /rustc/657bc01888e6297257655585f9c475a0801db6d2/library/std/src/sys/unix/thread.rs:71:17
  28:     0x7f8294dba609 - start_thread
  29:     0x7f8294cce293 - clone
  30:                0x0 - <unknown>

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.54.0-nightly (657bc0188 2021-05-31) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z verbose -C debuginfo=1 -C linker=/opt/compiler-explorer/gcc-11.1.0/bin/gcc --crate-type bin


```

</p>
</details>

### PR DESCRIPTION
**Pull request by oli-obk**
cc @lqd 

opaque types are not distinguishable from their hidden type at the codegen stage. So we could either end up with cases where the hidden type doesn't implement the trait (which will thus ICE) or where the hidden type does implement the trait (so we'd be using its impl instead of the one written for the opaque type). This can even lead to unsound behaviour without unsafe code.

Fixes https://github.com/rust-lang/rust/issues/86411.
Fixes https://github.com/rust-lang/rust/issues/84660.

rebase of #87382 plus some diagnostic tweaks

### CODE REVIEW

File: src/test/ui/type-alias-impl-trait/issue-84660-unsoundness.rs [hunk 1]
```diff
@@ -0,0 +1,41 @@
+// Another example from issue #84660, this time weaponized as a safe transmut: an opaque type in an
```
Comment (lqd) [inline src/test/ui/type-alias-impl-trait/issue-84660-unsoundness.rs:1]: tiny typo

```suggestion
// Another example from issue #84660, this time weaponized as a safe transmute: an opaque type in an
```

File: compiler/rustc_typeck/src/coherence/orphan.rs [hunk 2]
```diff
@@ -141,13 +142,56 @@ fn orphan_check_impl(tcx: TyCtxt<'_>, def_id: LocalDefId) -> Result<(), ErrorGua
...
+                hir::intravisit::walk_item(&mut visitor, item);
+                let reported = tcx
+                    .sess
+                    .struct_span_err(visitor.sp, "cannot implement trait on type alias impl trait")
+                    .span_note(tcx.def_span(def_id), "type alias impl trait defined here")
+                    .emit();
+                return Err(reported);
+            }
+        }
+        span_bug!(sp, "opque type not found, but `has_opaque_types` is set")
```
Comment (compiler-errors) [inline compiler/rustc_typeck/src/coherence/orphan.rs:194]:
```suggestion
        span_bug!(sp, "opaque type not found, but `has_opaque_types` is set")
```

**Review by compiler-errors — APPROVED** [review]
r=me with mine and lqd's typo nits fixed, or not (they're insignificant anyways)

**Pull request merged**